File: src/types/template.py
"""Pydantic schemas for template management."""
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from src.types.base import BaseSchema, IdentifiedModel, TimestampMixin
from src.types.enums import FailAction, FieldType, TemplateStatus


class FieldOptionSchema(BaseSchema):
    """Schema for a field option (select-type fields)."""

    id: uuid.UUID | None = None
    label: str
    value: str
    score: float = 0.0
    sort_order: int = 0


class TemplateFieldSchema(BaseSchema):
    """Schema for a template field."""

    id: uuid.UUID | None = None
    field_key: str
    label: str
    field_type: FieldType
    help_text: str | None = None
    is_mandatory: bool = False
    is_scoring: bool = False
    sort_order: int = 0
    options: list[FieldOptionSchema] = Field(default_factory=list)


class TemplateSectionSchema(BaseSchema):
    """Schema for a template section."""

    id: uuid.UUID | None = None
    name: str
    sort_order: int = 0
    fields: list[TemplateFieldSchema] = Field(default_factory=list)


class TemplateStageSchema(BaseSchema):
    """Schema for a template stage."""

    id: uuid.UUID | None = None
    name: str
    sort_order: int = 0
    weight_pct: float = 0.0
    min_pass_score: float | None = None
    fail_action: FailAction = FailAction.WARN
    sections: list[TemplateSectionSchema] = Field(default_factory=list)


class TemplateTagSchema(BaseSchema):
    """Schema for a template tag."""

    id: uuid.UUID | None = None
    tag: str


class TemplateCreate(BaseModel):
    """Payload for creating a template (Step 1)."""

    name: str = Field(..., min_length=1, max_length=255)
    category: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    icon: str | None = Field(None, max_length=50)
    theme_color: str | None = Field(None, max_length=7)
    tags: list[str] = Field(default_factory=list)


class TemplateUpdate(BaseModel):
    """Payload for updating template metadata."""

    name: str | None = Field(None, min_length=1, max_length=255)
    category: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    icon: str | None = Field(None, max_length=50)
    theme_color: str | None = Field(None, max_length=7)
    tags: list[str] | None = None


class StageInput(BaseModel):
    """Input for a single stage in Step 2."""

    name: str = Field(..., min_length=1, max_length=100)
    sort_order: int


class StagesUpdate(BaseModel):
    """Payload for updating stages (Step 2)."""

    stages: list[StageInput]


class FieldsUpdateStage(BaseModel):
    """Stage with sections/fields for Step 3 update."""

    stage_id: uuid.UUID
    sections: list[TemplateSectionSchema]


class FieldsUpdate(BaseModel):
    """Payload for bulk fields update (Step 3)."""

    stages: list[FieldsUpdateStage]


class ScoringStageInput(BaseModel):
    """Scoring config for a single stage (Step 4)."""

    stage_id: uuid.UUID
    weight_pct: float = Field(..., ge=0, le=100)
    min_pass_score: float | None = Field(None, ge=0, le=100)
    fail_action: FailAction = FailAction.WARN


class ScoringUpdate(BaseModel):
    """Payload for scoring update (Step 4)."""

    stages: list[ScoringStageInput]


class TemplateResponse(BaseSchema, TimestampMixin):
    """Template response with all nested data."""

    id: uuid.UUID
    name: str
    category: str
    description: str | None = None
    icon: str | None = None
    theme_color: str | None = None
    status: TemplateStatus
    version: int
    created_by: uuid.UUID
    tags: list[TemplateTagSchema] = Field(default_factory=list)
    stages: list[TemplateStageSchema] = Field(default_factory=list)


class TemplateListItem(BaseSchema):
    """Lightweight template for list views."""

    id: uuid.UUID
    name: str
    category: str
    description: str | None = None
    icon: str | None = None
    theme_color: str | None = None
    status: TemplateStatus
    version: int
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime
    stage_count: int = 0
    field_count: int = 0
    tags: list[str] = Field(default_factory=list)


class TemplateListResponse(BaseModel):
    """Paginated list of templates."""

    items: list[TemplateListItem]
    total: int
    page: int
    page_size: int

File: frontend/src/types/template.ts
export type TemplateStatus = 'draft' | 'published' | 'archived';
export type FieldType = 'text_short' | 'text_long' | 'single_select' | 'multi_select' | 'number' | 'date';
export type FailAction = 'warn' | 'block' | 'allow';

export interface FieldOption {
  id?: string;
  label: string;
  value: string;
  score: number;
  sort_order: number;
}

export interface TemplateField {
  id?: string;
  field_key: string;
  label: string;
  field_type: FieldType;
  help_text?: string;
  is_mandatory: boolean;
  is_scoring: boolean;
  sort_order: number;
  options: FieldOption[];
}

export interface TemplateSection {
  id?: string;
  name: string;
  sort_order: number;
  fields: TemplateField[];
}

export interface TemplateStage {
  id?: string;
  name: string;
  sort_order: number;
  weight_pct: number;
  min_pass_score: number | null;
  fail_action: FailAction;
  sections: TemplateSection[];
}

export interface TemplateTag {
  id?: string;
  tag: string;
}

export interface Template {
  id: string;
  name: string;
  category: string;
  description?: string;
  icon?: string;
  theme_color?: string;
  status: TemplateStatus;
  version: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  tags: TemplateTag[];
  stages: TemplateStage[];
}

export interface TemplateListItem {
  id: string;
  name: string;
  category: string;
  description?: string;
  icon?: string;
  theme_color?: string;
  status: TemplateStatus;
  version: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  stage_count: number;
  field_count: number;
  tags: string[];
}

export interface TemplateListResponse {
  items: TemplateListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface TemplateCreatePayload {
  name: string;
  category: string;
  description?: string;
  icon?: string;
  theme_color?: string;
  tags: string[];
}

export interface StageInput {
  name: string;
  sort_order: number;
}

export interface ScoringStageInput {
  stage_id: string;
  weight_pct: number;
  min_pass_score: number | null;
  fail_action: FailAction;
}

export interface FieldsUpdateStage {
  stage_id: string;
  sections: TemplateSection[];
}
